src/python/pants/rules/core/binary.py

```diff
@@ -62,13 +86,55 @@ class Binary(Goal):
 
 
 @rule
-async def coordinator_of_binaries(target: HydratedTarget) -> CreatedBinary:
-    binary = await Get[CreatedBinary](BinaryTarget, target.adaptor)
-    return binary
+async def coordinator_of_binaries(
+    wrapped_target: WrappedTarget,
+    union_membership: UnionMembership,
+    registered_target_types: RegisteredTargetTypes,
+) -> CreatedBinary:
+    target = wrapped_target.target
+    binary_implementations: Iterable[Type[BinaryImplementation]] = union_membership.union_rules[
+        BinaryImplementation
+    ]
+    valid_binary_implementations = [
+        binary_implementation
+        for binary_implementation in binary_implementations
+        if binary_implementation.is_valid_target(target)
+    ]
+    if not valid_binary_implementations:
+        all_valid_target_types = itertools.chain.from_iterable(
+            implementation.valid_target_types(
+                registered_target_types.types, union_membership=union_membership
+            )
+            for implementation in binary_implementations
+        )
+        formatted_target_types = sorted(target_type.alias for target_type in all_valid_target_types)
+        # TODO: this is a leaky abstraction that the error message knows this rule is being used
+        #  by `run` and `binary`. How should this be handled? A custom goal author could depend on
+        #  this error message too and the error would now be lying.
```

Comment:
I don't know what to do about this TODO. We could rewrite this error message to remove the leak but still be meaningful, e.g. "Cannot create a binary for this target type...".

Although, that's not as helpful or direct as "The run goal only works with ...". It's an implementation detail that `./v2 run` creates a binary to be able to then run that binary.

We could work around this by adding a new rule parameter with the error message so that call sites can customize the behavior. But, then `./v2 run` couldn't leverage the work of `./v2 binary` and vice versa.

Alternatively, we could somehow factor this `coordinator_of_binaries` rule out so that each call site can choose an error message that's meaningful to it? I'm not sure how trivial this would be.

--

Generally, I'd bias towards optimizing for the better error message and figuring out whatever gymnastics we need to support that great error message.

Reply:
Could the active goal name or even the full parse CLI be installed as a root rule so that rules wanting to know about that data could just request it?

Reply:
Yes, I believe at a minimum that we could have a singleton for all invoked goals. I don't think we could ask for the "active" goal because that requires accessing mutable global state.

Reply:
I feel like it would be reasonably foolproof to add the context in `run_goal_rules` above any raised exception... it'd be on two lines, but:
```
Exception: While executing the `run` goal:                                   # rendered by run_goal_rule
This implementation can only be used with the following target types:        # rendered by a rule
```

Passing down context is fine, but then you have to rely on rule implementers to actually use it.

Reply:
I think I like that! It's probably only worth rendering that message when multiple goals were given, though. If you only run `./v2 run`, that context wouldn't really help for anything.

Reply:
I think that it would probably still help in that case. It adds context that differentiates this from pants "just being broken" in some way, before getting to the goal.